fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryManager.java

```diff
@@ -121,18 +129,59 @@ public CompletableFuture<Void> mergeIndex(LucenePartitioner partitioner, LuceneA
         return cursor
                 .map(tuple ->
                         Tuple.fromItems(tuple.getItems().subList(0, groupingCount)))
-                .forEach(groupingKey ->
-                        mergeIndex(analyzerWrapper, groupingKey, partitioner.isPartitioningEnabled() ? 0 : null));
+                .forEachAsync(groupingKey -> mergeIndex(analyzerWrapper, groupingKey, partitioner, agilityContext),
+                        2);
     }
 
-    private void mergeIndex(LuceneAnalyzerWrapper analyzerWrapper, Tuple groupingKey, @Nullable Integer partitionId) {
-        try {
-            getDirectoryWrapper(groupingKey, partitionId, true).mergeIndex(analyzerWrapper);
-        } catch (IOException e) {
-            throw new RecordCoreStorageException("Lucene mergeIndex failed", e);
+    private CompletableFuture<Void> mergeIndex(LuceneAnalyzerWrapper analyzerWrapper, Tuple groupingKey,
+                                               @Nonnull LucenePartitioner partitioner, final AgilityContext agileContext) {
+        if (!partitioner.isPartitioningEnabled()) {
+            try {
+                getDirectoryWrapper(groupingKey, null, agileContext).mergeIndex(analyzerWrapper);
+                return AsyncUtil.DONE;
+            } catch (IOException e) {
+                throw new RecordCoreStorageException("Lucene mergeIndex failed", e)
+                        .addLogInfo(LuceneLogMessageKeys.GROUP, groupingKey);
+            }
+        } else {
+            AtomicReference<LucenePartitionInfoProto.LucenePartitionInfo> lastPartitionInfo = new AtomicReference<>();
+            return AsyncUtil.whileTrue(() -> getNextPartitionInfo(groupingKey, agileContext, lastPartitionInfo)
+                    .thenApply(partitionId -> mergePartition(analyzerWrapper, groupingKey, agileContext, partitionId, lastPartitionInfo)));
+        }
+    }
+
+    @Nonnull
+    private Boolean mergePartition(final LuceneAnalyzerWrapper analyzerWrapper, final Tuple groupingKey,
+                                   final AgilityContext agileContext, final Integer partitionId,
+                                   final AtomicReference<LucenePartitionInfoProto.LucenePartitionInfo> lastPartitionInfo) {
+        if (lastPartitionInfo.get() == null) {
```

Comment:
I'm not sure why there is an `AtomicReference` used here; I'll see if I can clean that up.